src/storage/src/source/mysql/replication.rs

```diff
@@ -0,0 +1,385 @@
+                        let received_upper = TransactionId::new(event.gno());
+                        assert!(new_upper <= received_upper, "GTID went backwards {} -> {}", new_upper, received_upper);
+                        new_upper = received_upper;
+
+                        // Indicate that we should advance the frontier if we receive a heartbeat since that would indicate
+                        // that there are no more records for this GTID.
+                        advance_on_heartbeat = true;
+                    }
+                    // A row event is a write/update/delete event
+                    Some(EventData::RowsEvent(data)) => {
+                        // Find the relevant table
+                        let binlog_table_id = data.table_id();
+                        let table_map_event = stream.get_ref().get_tme(binlog_table_id).ok_or_else(|| {
+                            TransientError::Generic(anyhow::anyhow!("Table map event not found"))
+                        })?;
+                        let table = match (
+                            table_id_map.get(&binlog_table_id),
+                            skipped_table_ids.get(&binlog_table_id),
+                        ) {
+                            (Some(table), None) => table,
+                            (None, Some(_)) => {
+                                // We've seen this table ID before and it was skipped
+                                continue;
+                            }
+                            (None, None) => {
+                                let table = table_name(&*table_map_event.database_name(), &*table_map_event.table_name())?;
+                                if table_info.contains_key(&table) {
+                                    table_id_map.insert(binlog_table_id, table);
+                                    &table_id_map[&binlog_table_id]
+                                } else {
+                                    skipped_table_ids.insert(binlog_table_id);
+                                    // We don't know about this table, so skip this event
+                                    continue;
+                                }
+                            }
+                            _ => unreachable!(),
+                        };
+
+                        let (output_index, table_desc) = &table_info[table];
+
+                        // Iterate over the rows in this RowsEvent. Each row is a pair of 'before_row', 'after_row',
+                        // to accomodate for updates and deletes (which include a before_row),
+                        // and updates and inserts (which inclued an after row).
+                        let mut rows_iter = data.rows(table_map_event);
+                        while let Some(Ok((before_row, after_row))) = rows_iter.next() {
+                            for (binlog_row, diff) in vec![(before_row, -1), (after_row, 1)].drain(..) {
```

Comment:
No need to allocate a vector here, you can use a static array. But also, I think a better way to express your intention here is the following, which will skip `None` values as part of the iterator

```rust
let updates = [before_row.map(|r| (r, -1), after_row.map(|r| (r, 1))];
for (inner_row, diff) in updates.into_iter().flatten() {
    // you get the inner row directly now, no need to match the option
}
```